analyzers/src/SonarAnalyzer.CSharp/Rules/AspNet/ControllersHaveTooManyResponsibilities.cs

```diff
@@ -0,0 +1,172 @@
+﻿/*
+ * SonarAnalyzer for .NET
+ * Copyright (C) 2015-2024 SonarSource SA
+ * mailto: contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System.Collections.Concurrent;
+
+namespace SonarAnalyzer.Rules.CSharp;
+
+[DiagnosticAnalyzer(LanguageNames.CSharp)]
+public sealed class ControllersHaveTooManyResponsibilities : SonarDiagnosticAnalyzer
+{
+    private const string DiagnosticId = "S6960";
+    private const string MessageFormat = "This controller has multiple responsibilities and could be split into {0} smaller units.";
+
+    private static readonly HashSet<string> ExcludedWellKnownServices =
```

Comment:
When a collection is used in a multi-threaded environment (as in multiple registrations), you need to have thread safety, so you need to use an immutable/frozen data structure, possibly with a lock-free behavior.
When a collection is used in a single-thread context, such as `RelevantMemberNames`, you don't need such guarantees.
I can add a comment to clarify that.

Reply:
I get the concept in theory, but aren't you using this HS in a multi-threaded context?
It is used in `SymbolStart`, which I assume happens concurrently for all the symbols.

Reply:
Yes, indeed. There is one difference, though: `ExcludedWellKnownServices` is instantiated statically, and it's never modified afterward. That's an important difference since it pretty much guarantees a memory barrier, which is important for the `Contains` to be de facto thread-safe (in practice, nothing is guaranteed, really, I am afraid—[ref](https://stackoverflow.com/a/28960891)).

Again in practice, if we couldn't assume read-safety of `HashSet<T>` and related, we would need to change a lot of code in our codebase.

On the other hand, `memberNames` is instantiated in `CheckApiController` and used in closures two levels down.
It is true we could use the same argument that `memberNames` is not modified after it's instantiated in `RelevantMemberNames`, so it should be safe as well. However, given the fact that the collection is local, and not static, I am a bit more reluctant to use a thread-unsafe collection here.

If you have arguments in favor, I can change `ImmutableHashSet<T>` to `HashSet<T>`.